// fats-command-release/lib/csslint.js
function csslintApp(filesArray, root, groupName, opt) {

	require('date-format-lite');

	var fs = require('fs'),
		path = require('path'),
		fats = require('fats-kernel'),
		csslintNode = require('csslint'),
		cssfile = filesArray,
		root = root.replace(/\\/g, '/');

	var buildfile = root + '/.report/report-csslint.html',  // 生成报告文件
		tpl = __dirname + '/report-template-csslint.html'; // 报告模板

	var errorLines = [],
		worker = null;

	//校验规则配置
	var ruleset = {
		'box-model': true,
		'display-property-grouping': true,
		'duplicate-properties': true,
		'empty-rules': true,
		'known-properties': true,
		'adjoining-classes': true,
		'box-sizing': true,
		'compatible-vendor-prefixes': true,
		'gradients': true,
		'text-indent': true,
		'vendor-prefix': true,
		'fallback-colors': true,
		'star-property-hack': true,
		'underscore-property-hack': true,
		'bulletproof-font-face': true,
		'font-faces': true,
		'import': true,
		'duplicate-background-images': true,
		'regex-selectors': true,
		'universal-selector': true,
		'unqualified-attributes': true,
		'zero-units': true,
		'overqualified-elements': true,
		'shorthand': true,
		'floats': true,
		'font-sizes': true,
		'ids': true,
		'important': true,
		'outline-none': true,
		'qualified-headings': true,
		'unique-headings': true
	};
	function getRules() {
		var _rule = fats.util.simpleCopy(ruleset);
		return fats.util.extend(_rule, opt);
	}

	/*
	 * Lint: lints css
	 */
	function lintCSS() {

		/*
		 * 遍历待检查的文件，生成结果数据
		 * */
		var data = [];
		for (var i = 0; i < cssfile.length; i++) {
			var css = fats.util.read(cssfile[i]),
				rules = getRules(),
				_temp;
			if (css.length > 0) {
				_temp = csslintNode.CSSLint.verify(css, rules);
				_temp['file'] = cssfile[i];
				data.push(_temp);
			}
		}

		/*
		 * 控制台输出概要信息
		 * */
		var needReport = false;
		var errorTotalCount = 0;
		for (var j = 0, len = data.length; j < len; j++) {
			var errorCount = 0,
				warningCount = 0,
				file = data[j].file.replace(root, ''),
				messages = data[j].messages;

			if (messages.length > 0) {
				needReport = true;
			}

			for (var m = 0, mlen = messages.length; m < mlen; m++) {
				if (messages[m].type == 'warning') {
					warningCount += 1;
				} else if (messages[m].type == 'error') {
					errorCount += 1;
					errorTotalCount += 1;
				}
			}
			console.log(' File ' + file.cyan + ' verified: ' +
				'error ' + errorCount.toString().red.bold + ',  ' +
				'warning ' + warningCount.toString().yellow.bold);

		}


		/*
		 * 根据模板生成校验报告
		 * */
		var reportDir = path.join(root, './.report');
		if (needReport) {

			/*
			 * 如果不存在，则建立 .report 文件夹
			 * */
			if (!fats.util.exists(reportDir)) {
				fs.mkdirSync(reportDir);
			}

			var template = fats.template;
			template.config('extname', '');

			var d = fats.util.localDate();
			var html = template(tpl, {
				data: data,
				version: fats.cli.info.version,
				time: d.format('UTC:YYYY-MM-DD hh:mm:ss')
			});
			// 写入报告文件
			fs.writeFileSync(buildfile, html);
			// 控制台提示报告已写入
			var _file = buildfile.replace(root, '');
			console.log(' Report file has been created: '.yellow.bold + _file.cyan.bold);
		} else {
			// 如果存在，则删除文件
			if (fats.util.exists(buildfile)) {
				fs.unlinkSync(buildfile);
			}
		}


		if (errorTotalCount > 0) {
			fats.log.error('Task has been terminated');
			process.exit(0);
		}
	}

	lintCSS();
}


module.exports = csslintApp;
